10401: evaluate Eq/Ne when the lhs - rhs evaluates

smichr opened the pull request:
The condition that the lhs and rhs be complex was added
without a test originally and it is not clear what it is
there for. The equality can still evaluate if one side
is real and the other complex.

The test that was corrected wouldn't fail as it was but
the inequality now evaluates, so the simpler expression
was used.

closes #10401 
closes #10687

smichr committed:
10401: evaluate Eq/Ne when the lhs - rhs evaluates

The condition that the lhs and rhs be complex was added
without a test originally and it is not clear what it is
there for. The equality can still evaluate if one side
is real and the other complex. The condition that both
sides be Expr guards against objects thinking they are
not zero (like MatAdd -- see issue #10728.

The test that was corrected wouldn't fail as it was, but
the inequality now evaluates so the simpler expression
was used.

smichr commented:
ping @mvnnn

mvnnn commented on sympy/core/tests/test_relational.py:
```diff
@@ -662,6 +662,17 @@ def test_issue_10304():
     assert simplify(Eq(e, 0)) is S.false
 
+
+def test_issue_10401():
```
need to add this test case for ```Eq```.
```
x = symbols('x', finite=False)
Eq((x - 1), x)
```
at``` x = oo```, ```(x-1) == x``` but currently it's return ```Eq(x - 1, x)``` so need to add condition for this in```class Equality```

smichr replied:
done as `Eq(inf1, inf2)` test.

mvnnn commented on sympy/core/tests/test_relational.py:
```diff
@@ -662,6 +662,17 @@ def test_issue_10304():
+    x = symbols('x', finite=False)
+    assert Ne(x/(x - 1), 1) is S.true
+    # the following is False if x == oo
+    assert Ne(x, x - 1).args == (x, x - 1)
```
this test case not useful for x = oo.
```
>>>x = symbols('x', finite=False)
>>> Ne(x, x - 1)
Ne(x, x - 1)
```
that's wrong output.

mvnnn replied:
we need this output
```
>>>x = symbols('x', finite=False)
>>> Ne(x, x - 1)
False
```
```
>>>x = symbols('x', finite=True)
>>> Ne(x, x - 1)
True
```
@smichr correct me if i am wrong.

smichr commented:
Let me reiterate and make sure we agree:

`Eq(x, x - 1)` is False if x is finite; if x is infinite perhaps we shouldn't decide. Let's see. If we have `Eq(f(x), g(x))` we could separate `f(x)` or `g(x)` into finite and infinite parts and, if `f` and `g` are Add's, ignore the finite parts. Let `F(x)` and `G(x)` be the infinite parts. If `F(x) == G(x)` then the equality is True. I'm not sure under what circumstances we can say that two infinite expressions are equivalent. Do they all go to `oo` and thus prove equal?

`Ne(x, x - 1)` is just the opposite.

And if the lhs is finite and the rhs infinite, the Eq is False.

smichr commented:
It's been updated @mvnnn (and now the case when lhs.is_finite != rhs.is_finite is handled).

mvnnn commented:
this test case still not work.
```
>>> x = Symbol('x', infinite=True)
>>> Ne(x/(x - 1), x)
Ne(x/(x - 1), x) <--- expected ans True
>>> Eq(x/(x - 1), x)
Eq(x/(x - 1), x)<---expected ans False
```

mvnnn commented on sympy/core/relational.py:
```diff
@@ -365,8 +372,17 @@ def __new__(cls, lhs, rhs, **options):
-            if is_equal == True or is_equal == False:
+            if isinstance(is_equal, BooleanAtom):
                 return ~is_equal
+            if 0:
```
remove this condition and make below code in else condition.

mvnnn commented:
this code return wrong ans 
```
>>> g = 1/(1 + 1/x)
>>> Eq(g, 1).subs(x, zoo)
False <--- expected ans True
>>> Ne(g, 1).subs(x, zoo)
True<---expected and False
```

smichr commented:
Now,

```
>>> x = Symbol('x', infinite=True)
>>> Ne(x/(x - 1), x)#<--- expected ans True
False
>>> Eq(x/(x - 1), x)#<---expected ans False
True
```

Those two failed because of the error in assumptions:

```
>>> (x/(x-1)).is_infinite
True
```

These are now as expected:

```
>>> g = 1/(1 + 1/x)
>>> Eq(g, 1).subs(x, zoo)#<--- expected ans True
True
>>> Ne(g, 1).subs(x, zoo)#<---expected and False
False
```

mvnnn commented:
```
>>> fin = symbols('inf', finite=True)
>>> inf = symbols('inf', infinite=True)
>>> zero = symbols('z', zero=True)
>>> Ne(zero*inf, inf)
False
>>> Ne(zero*fin, fin)
Ne(inf*z, inf)<--- expected ans True ("Sympy live Shell" gives ```>>> zero*fin == fin``` ans is ```False```)
```

mvnnn commented:
```
>>> a = Symbol("a")
>>> x = Symbol("x")
>>> Eq((x, a), (a, x))
Eq((x, a), (a, x)) <--- i think we need "None" ans. i am not sure
```

smichr commented:
> zero = symbols('z', zero=True)

any errors with args that are Expr should be handled with a new PR because I think the problem is elsewhere. In this case, `(zero*inf).is_finite` is incorrectly False. It should be NaN. This is an assumptions issue.

smichr commented:
> Eq((x, a), (a, x)) <--- expected ans "None" or "Error". i am not sure.

This, too, is a new issue. This PR is only handling issues that are related to the finite nature of the arguments.

smichr commented on sympy/core/tests/test_relational.py:
```diff
@@ -662,6 +662,33 @@ def test_issue_10304():
+    zero = symbols('z', zero=True)
+    nonzero = symbols('nz', zero=False, finite=True)
+    E = Eq
+    T, F = S.true, S.false
```
Eliminate the Ne tests since they are solved via Eq.

mvnnn replied:
you are right. we don't need same test case for both.

mvnnn replied:
@smichr not need to add same test cases for both because ```Ne``` also call ```Eq``` to check condition. need to (4-5) test cases for ```Eq``` and (4-5) test cases for ```Ne```(in both put different test cases and cover all below test cases).

smichr commented:
Except for new issues, are you ok with these changes @mvnnn?

mvnnn commented:
```
>>> p = Symbol('p', positive=True)
>>> Ne(p, 0)
Ne(p, 0)
```

smichr commented:
> Ne(p, 0)<--- expected ans True

Again, this is the new issue. And these cases make me think that this should be opened as a new issue...I'll do so now.

If the build passes after the recent mod I will commit what's here.

mvnnn commented:
@smichr that's test case already include in ```test_relational.py```. check this https://github.com/smichr/sympy/blob/rel/sympy/core/tests/test_relational.py#L94 and this https://github.com/smichr/sympy/blob/rel/sympy/core/tests/test_relational.py#L19.

mvnnn commented:
@smichr check my PR. i added condition for above test case and fix failed test cases.

smichr commented:
Leave a comment

smichr commented:
@mvnnn , I now see that the `is_zero` test is what was allowing some expressions to evaluate. I have updated that.

BTW, it gets a little confusing when multiple PRs with the same code modifications are opened. What I often do is make a PR against someone else's and let that be a place for discussions. Once modifications are done then that can be merged with the main PR and then the main PR can be committed.

smichr committed:
10728: Add.is_zero is None when noncommutative

smichr commented:
@mvnnn, are you ok with these changes?

mvnnn commented:
@smichr everything is fine except this.
```
>>> x = var('x')
>>> g = 1/(1 + 1/x)
>>> Eq(g, 1).subs(x, zoo) <--- expected and ```True```.
False
```
in both case expected ans is ```False```.(we also assume expected ans is ```None``` because ```oo/oo``` is not defined.)
```
>>> inf = symbols('inf', infinite=True)
>>> inf2 = symbols('inf2', infinite=True)
>>> Eq(inf/inf2,1)
False
>>> Eq(inf/inf,1)
True
```

smichr commented:
> g = 1/(1 + 1/x)

So I wonder what the test should be to not use `as_numer_denom` for such expressions? Do you have any ideas?

mvnnn commented on sympy/core/relational.py:
```diff
@@ -302,17 +302,37 @@ def __new__(cls, lhs, rhs=0, **options):
+                    if z is False and dif.is_commutative:  # issue 10728
+                        return S.false
+                    if z:
+                        return S.true
```
```
if any(isinstance(lhs, Pow) for (lhs, rhs)):
     pass
```
that's valid because normal Pow calculated in before condition.for example:
```
Eq(2**2, 4)
```
we can put this condition with below condition also.

smichr commented:
I modified without aid of compiler -- let's see how tests go. What do you think about the modification I made?

mvnnn commented:
@smichr you note that every time same test cases are fail from different files. so that the best way is to consider all those test cases and make condition. that's best way to cover most off condition.

smichr commented:
I'm not sure if this is what you mean, but this is what I'll try tonight:

```python
if n.is_finite:
    if n.is_zero:
        return d.is_nonzero
    elif n.is_zero is False:
        return d.is_infinite
elif n.is_infinite:
    return S.false
```

I've finally stepped back far enough to see the problem in general. Forgive me if you've been seeing this already and I have been slow to catch on. This, and the reversed Range work, have reinforced to me that some problems need strict enumeration to keep the code simpl(er).

mvnnn commented:
```
we need to check only infiniteness in else condition.
```
i mention this in my PR. that's ok. @smichr that's definitely works fine.update your PR.

smichr committed:
modify Eq conditions

I'm still not sure how to get something like Eq(exp(x)/(x-1), 1) to
evaluate to False if x is real. But perhaps this is beyond what should
be attempted in this routine.

smichr commented:
Please see not in last commit. Any ideas? Is there a simple answer that I am missing?

smichr commented:
How do the tests I've added fare in your PR?

mvnnn commented:
sorry for delay, currently i am busy to write GSOC proposal . we will solve this issue after 25th march.

mvnnn commented on sympy/core/tests/test_relational.py:
```diff
@@ -664,27 +664,24 @@ def test_issue_10304():
-    assert E(inf/fin, 0) is F
-    assert E(fin/inf, 0) is T
-    assert E(zero/nonzero, 0) is T and ((zero/nonzero) != 0)
+    assert Eq(fin, inf) is F
```
```
g = 1/(1 + 1/x)
assert Eq(g, 1).subs(x, zoo) is S.true
```
@smichr add this test case.

smichr replied:
I have now added it.

smichr replied:
I have now added it.

mvnnn replied:
+1

smichr committed:
add test showing why given Eq isn't false in general

smichr commented:
Do you have any other feedback, @mvnnn?

mvnnn commented:
+1

smichr commented:
Thanks for slogging through this with me. It's in.

cbm755 commented:
Late to this party...

> The condition that the lhs and rhs be complex was added
> without a test originally and it is not clear what it is

I think it was to keep noncommutative symbols out of there: (in hindsight the comment of mine in c87cf929cf2415c7e98b0128f2f338e44b5d81d5 was unclear: the "detect incompatibility" bit refers to use of `.is_zero`.).
